packages/content-conformance/src/engine.ts

```diff
@@ -0,0 +1,151 @@
+import fs from 'fs'
+import path from 'path'
+import { globbyStream } from 'globby'
+import type { Node } from 'unist'
+import { ContentFile } from './content-file.js'
+
+import type {
+  ConformanceRuleBase,
+  ConformanceRuleContext,
+  ContentConformanceFile,
+  DataFile,
+} from './types.js'
+import { ContentConformanceConfig } from './config.js'
+
+interface ContentConformanceEngineOptions
+  extends Omit<ContentConformanceConfig, 'rules'> {
+  rules: ConformanceRuleBase[]
+
+  /**
+   * Only load the specified files
+   */
+  files?: string[]
+}
+
+export class ContentConformanceEngine {
+  private opts: Omit<ContentConformanceEngineOptions, 'rules'>
+
+  private contentFiles: ContentFile[] = []
+
+  private dataFiles: DataFile[] = []
+
+  private rules: ConformanceRuleBase[] = []
+
+  constructor(opts: ContentConformanceEngineOptions) {
+    const { rules, ...restOpts } = opts
+    console.log(opts)
+
+    this.opts = restOpts
+    this.rules = opts.rules ?? []
+  }
+
+  async loadContentFiles() {
+    for await (const filepath of globbyStream(
+      this.opts.contentFileGlobPattern,
+      {
+        onlyFiles: true,
+        cwd: this.opts.root,
+      }
+    )) {
+      // If an array of filepaths are provided, only load the file if it matches one of the provided paths
+      if (
+        this.opts.files?.length &&
+        !this.opts.files.includes(String(filepath))
+      ) {
+        continue
+      }
+
+      const fullPath = path.join(this.opts.root, String(filepath))
+      const contents = await fs.promises.readFile(fullPath, 'utf-8')
+
+      const file = new ContentFile({
+        cwd: this.opts.root,
+        path: String(filepath),
+        value: contents,
+      })
+      this.contentFiles.push(file)
+    }
+  }
+
+  getContext(rule: ConformanceRuleBase): ConformanceRuleContext {
+    return {
+      contentFiles: this.contentFiles,
+      dataFiles: this.dataFiles,
+      report(message: string, file?: ContentConformanceFile, node?: Node) {
+        if (file) {
+          file.message(message, node, rule.id)
+        }
+      },
```

Comment:
💭 I'm curious how we can delineate between `warning` and `error`...

In discovery work, this is what I found to get the reporter to report and error instead of warning

```suggestion
      report(message: string, file?: ContentConformanceFile, node?: Node) {
        if (file) {
          let msg = file.message(message, node, rule.id)
          msg.fatal = true
        }
      },
```

<img width="541" alt="CleanShot 2022-11-09 at 11 32 36@2x" src="https://user-images.githubusercontent.com/26389321/200886888-12fc03fe-8bd0-4ec9-8dad-fc7a0307f03b.png">

Reply:
Yeah. I had initially not used the `fail()` method on the file as it immediately throws an error. I like setting `fatal` on the msg as you've done here though so we can re-use the existing reporter logic, as long as it doesn't immediately throw 😄  (I guess we could use `fail()` and just catch it immediately too)

Reply:
Related task: https://app.asana.com/0/1203331966445516/1203332374839504/f